🌱 Add logging guidelines (#6969)
kubernetes-sigs/cluster-api: docs/book/src/developer/logging.md

# Logging
The Cluster API project is committed to improving the SRE/developer experience when troubleshooting issues, and logging
plays an important part in this goal.

In Cluster API we strive to follow three principles while implementing logging:

- **Logs are for SRE & developers, not for end users!**
  Whenever an end user is required to read logs to understand what is happening in the system, most probably there is an
  opportunity for improvement of other observability in our API, like e.g. conditions and events.
- **Navigating logs should be easy**:
  We should make sure that SREs/Developers can easily drill down logs while investigating issues, e.g. by allowing to
  search all the log entries for a specific Machine object, eventually across different controllers/reconciler logs.
- **Cluster API developers MUST use logs!**
  As Cluster API contributors you are not only the ones that implement logs, but also the first users of them. Use it!
  Provide feedback!

## Upstream Alignment

> Kubernetes defines a set of [logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md), 

Author: https://github.com/kubernetes-sigs/cluster-api/pull/6969/files/91ccef7983b99f93bdaa30f8846f0c29e46dbbb3#diff-f519e7a9f2f1e96a5cbee2b82026a0de78a1836f889153e485e6581a7ad83e92 required to remove a similar note.
Also, we have specific pointers to Kubernetes guidelines all around the doc, so I don't think that having something super generic here adds much value.

> Kubernetes defines a set of [logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md), 

Reviewer: Should we just drop the `Upstream Alignment` section? I guess it should be fine since we are already referencing the upstream guidelines at various places in the document.

  the action log and the corresponding error log; While logging before the action, log verbs should use the -ing form.
- Ideally log messages should surface a different level of detail according to the target log level (see [log levels](#log-levels)
  for more details).

## Log Levels

Kubernetes provides a set of [recommendations](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use)
for log levels; as a small integration on the above guidelines we would like to add:

- Logs at the lower levels of verbosity (<=3) are meant to document “what happened” by describing how an object status
  is being changed by controller/reconcilers across subsequent reconciliations; as a rule of thumb, it is reasonable
  to assume that a person reading those logs has a deep knowledge of how the system works, but it should not be required
  for those persons to have knowledge of the codebase. 
- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
  the code; a person reading those logs usually has deep knowledge of the codebase. 
- Don’t use verbosity higher than 5.

> Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity for all the Cluster API
> controllers as recommended by the Kubernetes guidelines (currently it is 0).

Reviewer: ```suggestion
Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity (currently it is 0) for all the Cluster API
controllers as recommended by the Kubernetes guidelines.
```

This reads better IMHO. Up to you 🙂

# Logging
The Cluster API project is committed to improving the SRE/developer experience when troubleshooting issues, and logging
plays an important part in this goal.

In Cluster API we strive to follow three principles while implementing logging:

> - **Logs are for SRE & developers, not for users!**

Reviewer 2: We should define `users` in this context, SRE & developers can be users at the same time

Reviewer 2: Would you want to say application developer? Who's this persona?

Author: Switched to end user, defined in https://cluster-api.sigs.k8s.io/user/personas.html#service-provider-end-userconsumer

  Whenever an end user is required to read logs to understand what is happening in the system, most probably there is an
  opportunity for improvement of other observability in our API, like e.g. conditions and events.
- **Navigating logs should be easy**:
  We should make sure that SREs/Developers can easily drill down logs while investigating issues, e.g. by allowing to
  search all the log entries for a specific Machine object, eventually across different controllers/reconciler logs.

> - **Cluster API developers MUST use logs!**

Reviewer 2: ```suggestion
- **Cluster API developers MUST log!**
```

Author: it is intentionally must use

  As Cluster API contributors you are not only the ones that implement logs, but also the first users of them. Use it!
  Provide feedback!

## Upstream Alignment

Kubernetes defines a set of [logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md), 
as well as tools and libraries for logging.

As a CAPI project we must benefit from this great work, ensuring consistency with the ecosystem and focusing our

> resources on what is specific to this project.

Reviewer 2: Remove this? It's a bit hard to follow

Author: done

## Continuous improvement

While we all would like to have top-notch logs, we should recognize the fact that we already have a huge amount of code,
and changing logs is not the kind of work that attracts many contributors.

As a consequence we aim for continuous improvement, starting from a set of foundational items, that will enable a long

> tail of small improvements over time.

Reviewer 2: Remove these? Let's get right to the point of defining our logging

Author: done

The foundational items of Cluster API logging are:

> - Support for structured logging to all the Cluster API controllers (see [log format](#log-format)).

Reviewer 3: ```suggestion
- Support for structured logging in all the Cluster API controllers (see [log format](#log-format)).
```

- Using contextual logging (see [contextual logging](#contextual-logging)).
- Adding a minimal set of key/value pairs in the logger at the beginning of each reconcile loop, so all the subsequent
  log entries will inherit them (see [key value pairs](#keyvalue-pairs)).

> Starting from above foundations, then the long tail of small improvements will consist of following activities: 

Reviewer 3: ```suggestion
Starting from the above foundations, then the long tail of small improvements will consist of following activities: 
```

quoting the Kubernetes documentation, those are the key elements of this approach:

- Separate a log message from its arguments.
- Treat log arguments as key-value pairs.
- Be easily parsable and queryable.

Cluster API uses all the tooling provided by the Kubernetes community to implement structured logging: [Klog](https://github.com/kubernetes/klog), a
[logr](https://github.com/go-logr/logr) wrapper that works with controller runtime, and other utils for exposing flags
in the controller’s main.go.

Ideally, in a future release of Cluster API we will make JSON output format the default format for all the Cluster API
controllers (currently the default is still text format).

## Contextual logging

[Contextual logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging)
is the practice of using a log stored in the context across the entire chain of calls of a reconcile
action. One of the main advantages of this approach is that key value pairs which are added to the logger at the
beginning of the chain are then inherited by all the subsequent log entries created down the chain.

> Contextual is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's

Reviewer 3: ```suggestion
Contextual logging is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's
```

`LoggerFrom(ctx)` and `LoggerInto(ctx, log)` primitives and this ensures that:

> - The logger passed to each reconcile call has a unique `reconcileID`, so all the logs being writte during a single 

Reviewer 4: ```suggestion
- The logger passed to each reconcile call has a unique `reconcileID`, so all the logs being written during a single 
```

  for more details).

## Log Levels

Kubernetes provides a set of [recommendations](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use)
for log levels; as a small integration on the above guidelines we would like to add:

- Logs at the lower levels of verbosity (<=3) are meant to document “what happened” by describing how an object status
  is being changed by controller/reconcilers across subsequent reconciliations; as a rule of thumb, it is reasonable
  to assume that a person reading those logs has a deep knowledge of how the system works, but it should not be required
  for those persons to have knowledge of the codebase. 
- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
  the code; a person reading those logs usually has deep knowledge of the codebase. 
- Don’t use verbosity higher than 5.

Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity for all the Cluster API
controllers as recommended by the Kubernetes guidelines (currently it is 0).

## Anti patterns

> Developers should not use logs to try to reproduce the same level of details available in an IDE debugger.

Reviewer 2: Debatable? The limit of verbosity should be in the `--v` flag level, rather than limiting ourselves from the get-go. For example, a very high level might mean that we can allow tracing in some cases, which is close to an IDE debugger. You can't necessarily attach a debugger in production systems

Author: I have dropped this sentence and renamed the paragraph trade-offs
I agree that in some cases it could make sense to get closer to an IDE debugger.

readability vs machine readability, maintainability of the logs across the code base.

A reasonable approach for logging is to keep things simple and implement more log verbosity selectively and only on
thorny parts of code. Over time, based on feedback from SRE/developers, more logs can be added to shed light where necessary.

## Developing and testing logs

We are continuously improving our [Tilt](tilt.md) setup allowing Cluster API developers to use logs and improve them as part of
their development process, e.g. by allowing to deploy an entire log observability suite by simply setting the
`deploy_observability` value in your [tilt-setting.yaml](tilt.md#create-a-tilt-settings-file).

Over time, we would like to further improve this with the help of the community; feel free to open issues or propose
ideas!

## What about providers
Cluster API providers are developed by independent teams, and each team is free to define their own processes and
conventions.

However, given that SRE/developers looking at logs are often required to look both at logs from core CAPI and providers,

> we encourage the provider’s team to adopt and contribute to the guidelines defined in this document.

Reviewer 5: ```suggestion
we encourage providers to adopt and contribute to the guidelines defined in this document.
```
